tools/rust/keygen.rs

```diff
@@ -0,0 +1,59 @@
+// Copyright 2019 (c) rust-themis developers
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+use std::fs::OpenOptions;
+use std::io::{self, Write};
+#[cfg(unix)]
+use std::os::unix::fs::OpenOptionsExt;
+
+use clap::clap_app;
+use themis::keygen::gen_ec_key_pair;
+
+fn main() {
+    let matches = clap_app!(keygen =>
+        (version: env!("CARGO_PKG_VERSION"))
+        (about: "Generating ECDSA key pairs.")
+        (@arg secret: "Secret key file (default: key)")
+        (@arg public: "Public key file (default: key.pub)")
+    )
+    .get_matches();
+    let secret_path = matches.value_of("secret").unwrap_or("key").to_owned();
+    let public_path = matches
+        .value_of("public")
+        .map_or_else(|| secret_path.clone() + ".pub", |s| s.to_owned());
+
+    let (secret_key, public_key) = gen_ec_key_pair().split();
+
+    match write_file(&secret_key, &secret_path, 0o400) {
+        Ok(_) => {},
+        Err(e) => eprintln!("failed to write secret key to {}: {}", secret_path, e),
+    }
+    match write_file(&public_key, &public_path, 0o666) {
+        Ok(_) => {},
+        Err(e) => eprintln!("failed to write public key to {}: {}", public_path, e),
+    }
+}
+
+fn write_file(key: impl AsRef<[u8]>, path: &str, mode: u32) -> io::Result<()> {
+    let mut options = OpenOptions::new();
+    options.create(true);
+    options.truncate(true);
+    options.write(true);
+    #[cfg(unix)]
```

Comment:
just curious, what does it mean?

Reply:
That's how you do conditional compilation in Rust. The following expression will be compiled only on UNIX-like operating systems.

I saw that [all](https://github.com/cossacklabs/themis/blob/0d5ea505c23b4ecf786ef98d349765bf41d46a88/tools/go/keygen.go#L31) [other](https://github.com/cossacklabs/themis/blob/0d5ea505c23b4ecf786ef98d349765bf41d46a88/tools/python/keygen.py#L41) [wrappers](https://github.com/cossacklabs/themis/blob/master/tools/ruby/keygen.rb#L43) take care about file access modes for the key files so I did the same in Rust. However, `OpenOptions` does not have a `mode` method on all platforms, only on UNIX-like ones (well, can't blame Windows for not supporting UNIX access modes). It is imported only on UNIX as well:

https://github.com/cossacklabs/themis/blob/1d1268fdc78c85e6a49c7647808f9db8309de8aa/tools/rust/keygen.rs#L17-L18

We could technically omit `#[cfg(unix)]` if we're going to build Themis only on UNIX, but I believe in portable code.

Reply:
makes sense, thank you for the explanation (brilliant, as always!)